lib/libzfs/libzfs_import.c

```diff
@@ -94,30 +96,305 @@ typedef struct pool_list {
+	if (physpath != NULL && physpath[0] != '\0') {
+		(void) strlcpy(bufptr, physpath, buflen);
+		return (0);
+	}
+
+	return (ENODATA);
+}
+
+/*
+ * A disk is considered a multipath whole disk when:
+ *	DEVNAME key value has "dm-"
+ *	DM_NAME key value has "mpath" prefix
+ *	DM_UUID key exists
+ *	ID_PART_TABLE_TYPE key does not exist or is not gpt
+ */
+static boolean_t
+udev_mpath_whole_disk(struct udev_device *dev)
+{
+	const char *devname, *mapname, *type, *uuid;
+
+	devname = udev_device_get_property_value(dev, "DEVNAME");
+	mapname = udev_device_get_property_value(dev, "DM_NAME");
+	type = udev_device_get_property_value(dev, "ID_PART_TABLE_TYPE");
+	uuid = udev_device_get_property_value(dev, "DM_UUID");
+
+	if ((devname != NULL && strncmp(devname, "/dev/dm-", 8) == 0) &&
+	    (mapname != NULL && strncmp(mapname, "mpath", 5) == 0) &&
+	    ((type == NULL) || (strcmp(type, "gpt") != 0)) &&
+	    (uuid != NULL)) {
+		return (B_TRUE);
+	}
+
+	return (B_FALSE);
+}
+
+/*
+ * Check if a disk is effectively a multipath whole disk
+ */
+boolean_t
+is_mpath_whole_disk(const char *path)
+{
+	struct udev *udev;
+	struct udev_device *dev = NULL;
+	char nodepath[MAXPATHLEN];
+	char *sysname;
+	boolean_t wholedisk = B_FALSE;
+
+	if (realpath(path, nodepath) == NULL) {
+		return (B_FALSE);
+	}
+	sysname = strrchr(nodepath, '/') + 1;
+	if (strncmp(sysname, "dm-", 3) != 0) {
+		return (B_FALSE);
+	}
+	if ((udev = udev_new()) == NULL) {
+		return (B_FALSE);
+	}
+	if ((dev = udev_device_new_from_subsystem_sysname(udev, "block",
+	    sysname)) == NULL) {
+		udev_device_unref(dev);
+		return (B_FALSE);
+	}
+
+	wholedisk = udev_mpath_whole_disk(dev);
+
+	udev_device_unref(dev);
+	return (wholedisk);
+}
+
+static int
+udev_device_is_ready(struct udev_device *dev)
+{
+#ifdef HAVE_LIBUDEV_UDEV_DEVICE_GET_IS_INITIALIZED
+	return (udev_device_get_is_initialized(dev));
+#else
+	/* wait for DEVLINKS property to be initialized */
+	return (udev_device_get_property_value(dev, "DEVLINKS") != NULL);
+#endif
+}
+
+/*
+ * Encode the persistent devices strings
+ * used for the vdev disk label
+ */
+static int
+encode_device_strings(const char *path, vdev_dev_strs_t *ds,
+    boolean_t wholedisk)
+{
+	struct udev *udev;
+	struct udev_device *dev = NULL;
+	char nodepath[MAXPATHLEN];
+	char *sysname;
+	int ret = ENODEV;
+	int sleepcnt = 0;
+
+	if ((udev = udev_new()) == NULL) {
+		return (ENXIO);
+	}
+
+	/* resolve path to a runtime device node instance */
+	if (realpath(path, nodepath) == NULL) {
+		goto no_dev;
+	}
+	sysname = strrchr(nodepath, '/') + 1;
+
+	/*
+	 * Wait up to 3 seconds for udev to set up the device node context
+	 */
+	while (sleepcnt < 300) {
+		if ((dev = udev_device_new_from_subsystem_sysname(udev,
+		    "block", sysname)) == NULL) {
+			goto no_dev;
+		}
+		if (udev_device_is_ready(dev)) {
+			break;  /* udev ready */
+		}
+		udev_device_unref(dev);
+		dev = NULL;
+
+		(void) usleep(10000);	/* 10 ms */
```

Comment:
@ryao I did look at using a udev monitor (we use it in upcoming ZED work). However it would cause significant refactoring. Ideally the `zpool_label_disk_wait()` would be re-worked to be based on udev monitoring instead of polling and would eliminate the need for `update_vdev_config_dev_strs()` to wait at all.